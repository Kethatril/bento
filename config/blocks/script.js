module.exports = function(config, options) {
    config.merge({
        module: {
            rules: [
                {
                    test: /\.js$/,
                    exclude: /(node_modules)/,
                    use: {
                        loader: "babel-loader",
                        options: options.get("babel")
                    }
                }
            ]
        }
    });
};
