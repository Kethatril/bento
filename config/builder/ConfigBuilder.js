const { Config } = require("webpack-config"),
    parse = require("./modules/parse");

module.exports = class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    constructor(options = {}) {
        this._options = parse(options);
        this._config = new Config();
        this._modules = [];
        this._manifest = {};
    }

    use(...modules) {
        this._modules = this._modules.concat(modules);
        return this;
    }

    build(fn) {
        this._modules.push(fn);

        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._modules,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            let config = new Config(),
                options = this._options;

            try {
                (function next(err) {
                    let async;

                    // Increment counter variable and skip any indices that
                    // don't exist. This allows sparse arrays to be iterated.
                    do {
                        ++i;
                    } while (!(i in modules) && i !== len);

                    if (err === false || err) return reject(err);

                    if (i === len) return resolve(config.toObject());

                    let result;

                    result = modules[i].call(
                        {
                            async() {
                                async = true;
                                return next;
                            }
                        },
                        config,
                        options
                    );
                    if (!async) {
                        next(result);
                    }
                })();
            } catch (e) {
                reject(e);
            }
        });
    }
};
