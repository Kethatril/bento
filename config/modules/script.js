const { config, read } = require("../utils");

module.exports = config(instance => {
    return instance.merge({
        module: {
            rules: [
                {
                    test: /\.js$/,
                    exclude: /(node_modules)/,
                    use: {
                        loader: "babel-loader",
                        options: read("babel")
                    }
                }
            ]
        }
    });
});
