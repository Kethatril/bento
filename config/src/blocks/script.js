import merge from "webpack-merge";
import webpack from "webpack";

export default function script(config, options) {
    config = merge(config, {
        module: {
            rules: [
                {
                    test: /\.js$/,
                    exclude: /(node_modules)/,
                    use: {
                        loader: "babel-loader",
                        options: options.get("babel")
                    }
                }
            ]
        }
    });

    if (options.get("env.isProduction")) {
        config = merge(config, {
            plugins: [
                new webpack.optimize.CommonsChunkPlugin({
                    name: "_manifest"
                }),

                new webpack.optimize.CommonsChunkPlugin({
                    name: "_vendor",
                    minChunks: function(module) {
                        return /node_modules/.test(module.resource);
                    }
                }),

                new webpack.optimize.UglifyJsPlugin({
                    sourceMap: options.get("enabled.sourceMap")
                })
            ]
        });
    }

    return config;
}
