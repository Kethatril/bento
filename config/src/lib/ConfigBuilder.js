import { parse, resolve, select } from "~/lib/utils";
import ManifestKeeper from "~/lib/ManifestKeeper";

export default class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    manifest = new ManifestKeeper();

    constructor(options = {}) {
        this._options = parse(resolve(options));
        this._blocks = options.blocks.list || [];
        this._timeout = options.blocks.timeout || 3000;
    }

    get options() {
        return this._options;
    }

    build() {
        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._blocks,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            const timeout = this._timeout;

            let config = {},
                options = this._options,
                utils = { select: select(options), manifest: this.manifest };

            (function next(result) {
                let async;

                // Increment counter variable and skip any indices that
                // don't exist. This allows sparse arrays to be iterated.
                do {
                    ++i;
                } while (!(i in modules) && i !== len);

                if (result === false || result instanceof Error)
                    return reject(err);

                if (i === len) {
                    return resolve(result);
                }

                try {
                    result = modules[i](result, options, utils);
                } catch (e) {
                    return reject(e);
                }
                if (!result || typeof result !== "object") {
                    throw new Error(
                        "[InvalidValueError] Each block must either return a new config object or a promise that resolves a new config object."
                    );
                }

                if (result.then) {
                    const id = setTimeout(function() {
                        if (!async) return;
                        return reject(
                            new Error(
                                `[TimeoutError] ${timeout} ms exceeded. For async blocks, ensure the returned promise is resolved.`
                            )
                        );
                    }, timeout);

                    const done = function(result) {
                        clearTimeout(id);
                        next(result);
                    };

                    async = true;
                    result.then(done, done);
                }

                if (!async) {
                    next(result);
                }
            })(config);
        });
    }
}
