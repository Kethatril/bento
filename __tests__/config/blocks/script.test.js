import script from "~/blocks/script";
import ConfigBuilder from "~/lib/ConfigBuilder";

describe("blocks/script", () => {
    const options = {
        babel: {},
        blocks: {
            list: [script]
        }
    };

    it("matches snapshot when in development", () => {
        const builder = ConfigBuilder.create(
            Object.assign(
                {},
                {
                    env: {
                        isProduction: false
                    }
                },
                options
            )
        );

        return builder.build().then(config => {
            expect(config).toMatchSnapshot();
        });
    });

    it("matches snapshot when in production", () => {
        const builder = ConfigBuilder.create(
            Object.assign(
                {},
                {
                    env: {
                        isProduction: true
                    }
                },
                options
            )
        );

        return builder.build().then(config => {
            expect(config).toMatchSnapshot();
        });
    });
});
