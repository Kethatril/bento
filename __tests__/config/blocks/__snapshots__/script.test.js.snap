// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blocks/script matches snapshot when in development 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /(node_modules)/,
        "test": /\\.js$/,
        "use": Object {
          "loader": "babel-loader",
          "options": Object {},
        },
      },
    ],
  },
}
`;

exports[`blocks/script matches snapshot when in production 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /(node_modules)/,
        "test": /\\.js$/,
        "use": Object {
          "loader": "babel-loader",
          "options": Object {},
        },
      },
    ],
  },
  "plugins": Array [
    CommonsChunkPlugin {
      "async": undefined,
      "children": undefined,
      "chunkNames": Array [
        "_vendor",
      ],
      "filenameTemplate": undefined,
      "ident": "<cwd>/node_modules/webpack/lib/optimize/CommonsChunkPlugin.js0",
      "minChunks": [Function minChunks],
      "minSize": undefined,
      "selectedChunks": undefined,
    },
    CommonsChunkPlugin {
      "async": undefined,
      "children": undefined,
      "chunkNames": Array [
        "_manifest",
      ],
      "filenameTemplate": undefined,
      "ident": "<cwd>/node_modules/webpack/lib/optimize/CommonsChunkPlugin.js1",
      "minChunks": undefined,
      "minSize": undefined,
      "selectedChunks": undefined,
    },
    UglifyJsPlugin {
      "options": Object {
        "sourceMap": undefined,
      },
    },
  ],
}
`;
